Use PATCHs instead of PUTs in CIDR allocator

shyamjvs opened the pull request:
Ref https://github.com/kubernetes/kubernetes/issues/52292

/cc @wojtek-t 

cc @kubernetes/sig-network-misc @bowei @MrHohn 

```release-note
NONE
```

shyamjvs committed:
Use PATCHs instead of PUTs in CIDR allocator

shyamjvs commented:
/kind bug
/priority critical-urgent
/sig network

Getting this into 1.9 for performance improvement of cidr-allocator.

shyamjvs commented:
Note: I might have to change the unit tests to mock the new behavior in fakeNodeHandler.

xiangpengzhao commented:
/test pull-kubernetes-e2e-gce

wojtek-t commented on pkg/controller/node/ipam/cloud_cidr_allocator.go:
```diff
@@ -187,8 +187,7 @@ func (ca *cloudCIDRAllocator) updateCIDRAllocation(nodeName string) error {
 		}
-		node.Spec.PodCIDR = podCIDR
-		if _, err = ca.client.CoreV1().Nodes().Update(node); err == nil {
+		if err = nodeutil.PatchNodeCIDR(ca.client, types.NodeName(node.Name), podCIDR); err == nil {
```
Please remove the TODO from line 171

shyamjvs replied:
Nice catch. Done.

wojtek-t commented on pkg/controller/node/ipam/range_allocator.go:
```diff
@@ -248,8 +250,7 @@ func (r *rangeAllocator) updateCIDRAllocation(data nodeAndCIDR) error {
 		}
-		node.Spec.PodCIDR = podCIDR
-		if _, err = r.client.CoreV1().Nodes().Update(node); err == nil {
+		if err = nodeutil.PatchNodeCIDR(r.client, types.NodeName(node.Name), podCIDR); err == nil {
```
Please remove the TODO from line 236

shyamjvs replied:
Done.

wojtek-t commented:
/lgtm

wojtek-t commented:
/approve

shyamjvs commented:
I've fixed a couple of bugs due to which the presubmits failed:
- I needed to json encode even the cidr (though it's a string, as json follows utf-8). Due to this range-allocator was failing and those clusters were failing to startup.
- There is an integration test called "Correctly recycle CIDR" for the range-allocator in which we first allocate a CIDR to a node, then release it, then again try ask for a CIDR for the node and check that the one we get is same as the earlier one. Turns out this test was wrong and earlier it was passing due to a bug in the test. As the new CIDR allocated should actually be the next one to the previously allocated CIDR (even though that one got released). This seems to be the way cidr-set has been implemented (where we're using bitmap and cyclically going around marking the next unused one in the chain). I changed the test to the following now:

A (127.123.234.0/30), B (127.123.234.4/30), C (127.123.234.8/30), D (127.123.234.12/30) are the available CIDRs
I marked B, C and D as already pre-allocated (so they're not free)
Then A gets assigned to the node
We're then releasing A.
Then we're requesting CIDR for the node again... and we're getting A again (previously it would've been B).

shyamjvs commented:
@wojtek-t Could you please re-lgtm?

wojtek-t commented:
/lgtm
